keras_core/trainers/trainer.py

```diff
@@ -121,9 +121,12 @@ def run_eagerly(self, value):
 
     @property
     def metrics(self):
-        metrics = [self._loss_tracker]
+        if hasattr(self, "_loss_tracker"):
+            metrics = [self._loss_tracker]
+        else:
+            metrics = []
         metrics.extend(self._metrics[:])
-        if self._compile_metrics is not None:
+        if getattr(self, "_compile_metrics", None) is not None:
```

Comment:
Why this change?

- The loss tracker should already be part of `metrics`
- `metrics` should definitely be called after `__init__` so `_compile_metrics` should exist (if not that seems like a bug)

Reply:
I was hitting some test error for this. The `loss_tracker` is only inited after model.compile(), same for `_compile_metrics `.

This was causing issue when I try to access model.metric_variables for a uncompiled model, eg during model.predict().

Reply:
See test error at https://github.com/keras-team/keras-core/actions/runs/6054618513/job/16432299542.

Reply:
I think you should be able to fix it by adding a `if self.compiled:` clause in `_record_training_state_sharding_spec`/etc to only look at optimizer variables and metric variables if the model is compiled, and not before.

Reply:
Oh, I think in this case, the `if self.compiled` should be added here, since `_loss_tracker` and `_compile_metrics` will only be populated after model.compile. Also the fix has to happen here, otherwise for a unacompiled model/layer, the model.metrics will raise an error which is a bad UX.